### ISSUE TITLE
substantially rate limit unauthenticated /api/ queries to encourage authenticated queries

### ISSUE THREAD

**Issue by ccaputo** [peeringdb/peeringdb#853]
There is heavy /api/ usage by unauthenticated clients.  This is a problem for Ops, because it is difficult to figure out who is behind the queries.  That is needed when the user behind the queries needs to be contacted due to broken scripts, poorly designed scripts, or abuse.

Once https://github.com/peeringdb/peeringdb/issues/427 is implemented, I recommend unauthenticated /api/ queries be substantially rate limited so as to promote authenticated usage.

**Comment by ynbrthr** [comment]
+1

**Comment by arnoldnipper** [comment]
> There is heavy /api/ usage by unauthenticated clients

I wouldn't call it heavy. However, it's true that ~95% of all API calls are unauthenticated. Which is fine per se. We want to enable rate-limiting to prevent the infrastructure from DDoS or broken setups, but not to discourage users to retrieve data.

On the other hand, @oc should closely monitor the infrastructure and scale it when it needed.

**Comment by ccaputo** [comment]
> > There is heavy /api/ usage by unauthenticated clients
> 
> I wouldn't call it heavy. However, it's true that ~95% of all API calls are unauthenticated. Which is fine per se. We want to enable rate-limiting to prevent the infrastructure from DDoS or broken setups, but not to discourage users to retrieve data.
> 
> On the other hand, @peeringdb/oc should closely monitor the infrastructure and scale it when it needed.

I am advising that the status quo makes it challenging for the @peeringdb/oc to contact problem sources.  If we evolve toward a future in which most queries are authenticated, that will be helpful.  The alternative is to only have an IP address to work from.  With cloud services, that makes reaching the end-user very difficult, leaving just the alternative of blocking problem source IP addresses.

**Comment by ccaputo** [comment]
I'd recommend this happen _after_ https://github.com/peeringdb/peeringdb/issues/266 (add API keys) has been implemented, since that provides a secure way to have credentials on automated clients.

**Comment by arnoldnipper** [comment]
Demanding an account is imho perpendicular to preaching that almost all of the data is accessible publicly. That would mean that we have a 180° change in our paradigm. Hence, steering towards demanding authenticated API calls ist the wrong approach imho.

**Comment by job** [comment]
@arnoldnipper it is quite common to rate limit unauthenticated API endpoints. For example, the RIPE database by default rate limits specific WHOIS queries, unless your IP is on a special allowlist. The data is and remains accessible to the public, but _hammering_ without identifying yourself probably should be (rate) limited.

**Comment by arnoldnipper** [comment]
@job, I'm not advocating to not rate-limiting. However, $subject suggests something different!

**Comment by ccaputo** [comment]
> @job, I'm not advocating to not rate-limiting. However, $subject suggests something different!

@arnoldnipper, is the new subject good?

**Comment by arnoldnipper** [comment]
no, imho everything should and could be handled in #427. Spreading one topic across several issues does not really help. Adjusting subject in the other issue might help, though

**Comment by ccaputo** [comment]
> no, imho everything should and could be handled in #427. Spreading one topic across several issues does not really help. Adjusting subject in the other issue might help, though

Ok, I reverted the subject.  #427 does not address the issue in my opinion, it just provides the ability.  That is why this issue depends on #427 and is policy oriented.

Further, we should generally be encouraging authenticated use of the API in our documentation/presentations.

**Comment by arnoldnipper** [comment]
I still think we should/could handle everything in #427 instead of spreading it across a couple of issues

**Comment by ynbrthr** [comment]
@peeringdb/pc , we need a few +1s

**Comment by shane-kerr** [comment]
+1

I have no preference for whether this happens before, after, or inside of #427.

**Comment by arnoldnipper** [comment]
-1 here and do it in one go in #427

**Comment by grizz** [comment]
Technically, this is just the first part of #427 -- https://github.com/peeringdb/peeringdb/issues/427#issuecomment-650574257

For policy, I would think the numbers are up to ops, and ops should work with the OC to encourage authenticated usage.

**Comment by arnoldnipper** [comment]
> For policy, I would think the numbers are up to ops, and ops should work with the OC to encourage authenticated usage.

What we've seen was an insane number of requests. These numbers also don't make sense for an authenticated user. We have had that mantra that you don't need to register with PeeringDB to use it. And with my @peeringdb/outcom member hat on, we should retain this openness.

**Comment by arnoldnipper** [comment]
I'm changing my -1 to a 0

**Comment by vegu** [comment]
If i understand correctly this would just a be config change to `API_THROTTLE_RATE_ANON` (currently set to 100/sec)

**Comment by ccaputo** [comment]
> If i understand correctly this would just a be config change to `API_THROTTLE_RATE_ANON` (currently set to 100/sec)

From an @peeringdb/oc committee member perspective, I would love to see both API_THROTTLE_RATE_ANON and API_THROTTLE_RATE_USER be able to be viewed/set in the /cp/ by Ops so that a restart of daemons/containers is not needed for any changes to take effect.

Ops can then adjust these based on observation and testing as the back-end evolves.
